### PR TITLE
DAOS-5465 csum: Corrupt BIO Address through VOS

### PR DESCRIPTION
**Pull request by ryon-jensen**
Add the ability to set a bio address as corrupted using the VOS
iter API. This will be necessary for the checksum scrubber
to be able to mark data as corrupted when silent data
corruption is detected. Change the vos_iter_delete api to be 
more generic with an op code parameter. Will have a delete 
and corrupt op code for now.

Signed-off-by: Ryon Jensen <ryon.jensen@intel.com>

### CODE REVIEW

File: src/vos/vos_tree.c [hunk 1]
```diff
@@ -779,6 +796,7 @@ static btr_ops_t singv_btr_ops = {
 	.to_rec_free		= svt_rec_free,
 	.to_rec_fetch		= svt_rec_fetch,
 	.to_rec_update		= svt_rec_update,
+	.to_rec_corrupt		= svt_rec_corrupt,
```
Comment (jolivier23) [inline src/vos/vos_tree.c:799]: can we not do this through rec_update for single value?   It's only needed for single value, right?

File: src/vos/vos_obj.c [hunk 2]
```diff
@@ -1866,23 +1887,33 @@ vos_obj_iter_aggregate(daos_handle_t ih, bool discard)
...
+	switch (op) {
+	case VOS_ITER_PROC_OP_DELETE:
+		switch (iter->it_type) {
+		default:
+			D_ASSERT(0);
+			return -DER_INVAL;
+		case VOS_ITER_DKEY:
+		case VOS_ITER_AKEY:
+		case VOS_ITER_SINGLE:
+			return obj_iter_delete(oiter, args);
```
Comment (jolivier23) [inline src/vos/vos_obj.c:1904]: I think we should just call an API here that is specific to single value rather than hooking up the corrupt callbacks and such

**Review by jolivier23 — CHANGES_REQUESTED** [review]

File: src/vos/vos_tree.c (hunk 1 above)
Comment (NiuYawei) [reply src/vos/vos_tree.c:799]: agreed, looks it's not a common thing, could be done via svt_rec_update()?

Comment (ryon-jensen) [reply src/vos/vos_tree.c:799]: Well, I tried that, but it "/** Disallow[s] same epoch overwrite */", and then appears to frees & allocate mem.

File: src/vos/vos_obj.c (hunk 2 above)
Comment (ryon-jensen) [reply src/vos/vos_obj.c:1904]: Suggestions? I don't see any single value APIs available at the vos object level ... unless we introduce a new api layer? Arrays have the api from evtree.h, but the btree apis are only generic (not knowing that it's a single value), right?

File: src/vos/vos_internal.h [hunk 3]
```diff
@@ -887,8 +887,8 @@ struct vos_iter_ops {
 	int	(*iop_copy)(struct vos_iterator *iter,
 			    vos_iter_entry_t *it_entry, d_iov_t *iov_out);
 	/** Delete the record that the cursor points to */
-	int	(*iop_delete)(struct vos_iterator *iter,
-			      void *args);
+	int	(*iop_process)(struct vos_iterator *iter, vos_iter_proc_op op,
```
Comment (Nasf-Fan) [inline src/vos/vos_internal.h:890]: Compared with PR#5501, this seems more general solution. Currently, we set a DTX entry as 'corrupted' via vos_dtx_abort_one(). I prefer to unify them via the same API if possible. This patch offers some way for that.

**Review by jolivier23 — CHANGES_REQUESTED** [review]
We discussed a solution to the btree callback issue I previously mentioned so requesting changes to get it off my list.

File: src/include/daos/btree.h [hunk 4]
```diff
@@ -343,6 +343,7 @@ typedef struct {
 	int		(*to_rec_update)(struct btr_instance *tins,
 					 struct btr_record *rec,
 					 d_iov_t *key, d_iov_t *val);
+
```
Comment (jolivier23) [inline src/include/daos/btree.h:346]: Minor, revert this change

**Review by jolivier23 — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

**Review by Nasf-Fan — CHANGES_REQUESTED** [review]
Do we need some mechanism to mark the dkey/akey as 'corrupted' if the checksum for related key is failed?

**Comment by ryon-jensen** [comment]
> Do we need some mechanism to mark the dkey/akey as 'corrupted' if the checksum for related key is failed?

@Nasf-Fan - I don't think so because the key checksums aren't stored. They are calculated for every I/O.

File: src/vos/vos_io.c [hunk 5]
```diff
@@ -714,6 +715,11 @@ akey_fetch_single(daos_handle_t toh, const daos_epoch_range_t *epr,
 	if (ci_is_valid(&csum_info))
 		save_csum(ioc, &csum_info, NULL, 0);
 
+	if (BIO_ADDR_IS_CORRUPTED(&rbund.rb_biov->bi_addr)) {
+		D_DEBUG(DB_CSUM, "Found corrupted record\n");
+		return -DER_CSUM;
```
[the remark this answers is not in the document]
Comment (ryon-jensen) [reply src/vos/vos_io.c:720]: Currently I can't think of other ways to detect data corruption that would set the bio_addr to corrupted. Going to leave as -DER_CSUM for now, but if we do add other ways to set bio_addr to corrupted, we can change in the future.

File: src/vos/vos_obj.c [hunk 6]
```diff
@@ -1866,23 +1914,33 @@ vos_obj_iter_aggregate(daos_handle_t ih, bool discard)
...
+		case VOS_ITER_RECX:
+			return evt_iter_delete(oiter->it_hdl, NULL);
+		}
+	case VOS_ITER_PROC_OP_MARK_CORRUPT:
+		if (iter->it_type == VOS_ITER_SINGLE)
+			return sv_iter_corrupt(oiter);
+		else if (iter->it_type == VOS_ITER_RECX)
+			return evt_iter_corrupt(oiter->it_hdl);
```
[the remark this answers is not in the document]
Comment (ryon-jensen) [reply src/vos/vos_obj.c:1939]: I agree. Was just copying what was already there (obj|evt_iter_delete...). I will update.

**Review by jolivier23 — APPROVED** [review]

File: src/vos/vos_obj.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (Nasf-Fan) [reply src/vos/vos_obj.c:1945]: In fact, what we need for obj_iter_corrupt() is not the "vos object from the iterator", instead, it is the "umem_instance" for PMDK TX. That is also needed for evt_iter_corrupt(). The difference for evt_iter_corrupt() is that: evt_iter_corrupt() can obtain the "umem_instance" from the evt_context. I am not mind to explicitly pass the "*umm" parameter if you want to unify them.

**Review by Nasf-Fan — APPROVED** [review]

File: doc/overview/data_integrity.md [hunk 7]
```diff
@@ -385,7 +385,7 @@ obj_iter_scrub(coh, epr, csummer, pool_uuid, event_handlers, entry, type)
   includes a CORRUPTED bit.
 - The vos update api already accepts a flag, so a CORRUPTED flag is added and
   handled during an update so that, if set, the bio address will be updated to
-  be corrupted.
+  be corrupted. (Implemented by DAOS-5465 and documented in src/vos/README.md)
```
[the remark this answers is not in the document]
Comment (ryon-jensen) [reply doc/overview/data_integrity.md:388]: Probably not. Will remove.

**Comment by ryon-jensen** [comment]
Closing this PR and will open a new one to be landed on feature/csum_scrubbing
